Fixed custom Exception classes

ronaldbarendse opened the pull request:
### Prerequisites
If there's an existing issue for this PR then this fixes https://github.com/umbraco/Umbraco-CMS/issues/5947

### Description
All custom `Exception` classes shoud be updated with:
- `[Serializable]` attribute (as any class that might be serialized must be marked with this attribute)
- Constructor with the signature `(SerializationInfo information, StreamingContext context)` (as required for all classes that implement `ISerializable`)
- If custom properties are added, restore the data in the above constructor and override `GetObjectData(SerializationInfo info, StreamingContext context)` to save the data

ronaldbarendse commented on src/Umbraco.Core/Exceptions/ArgumentNullOrEmptyException.cs:
```diff
@@ -1,30 +1,54 @@
     /// </summary>
+    /// <seealso cref="System.ArgumentNullException" />
+    [Serializable]
     public class ArgumentNullOrEmptyException : ArgumentNullException
```
I see this exception is used quite a lot, but it's semantically not correct. If this is thrown for empty values, the base class is `ArgumentNullException`, but the argument isn't `null`... And as the current value isn't available nor the message specific enough, debugging whether it was `null` or empty will be unnecessarily hard.

Yes, the following code is quite cumbersome to write, but it's correct and gives good debug information:
```c#
if (param == null) throw new ArgumentNullException(nameof(param));
if (string.IsNullOrWhiteSpace(param)) throw new ArgumentException("Value can't be empty or consist only of white-space characters.", nameof(param));
```

See also:
- https://docs.microsoft.com/en-us/dotnet/api/system.argumentexception (explains the differences between the exceptions)
- https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/code-contracts (might be a better way to do these validations)

Shazwazza replied:
I wasn't aware of this exception or why it exists. Both Visual Studio and Resharper offer a very easy way to check strings for both null and empty by just pressing ctrl + enter. I personally don't see a reason for this exception but i don't see a reason to throw both an ArgumentNullException and an ArgumentException.

Shazwazza replied:
Visual Studio and Resharper both generate this single check for null/empty strings:

ronaldbarendse replied:
Well, the linked documentation from Microsoft clearly states:

> The primary derived classes of `ArgumentException` are `ArgumentNullException` and `ArgumentOutOfRangeException`. These derived classes should be used instead of `ArgumentException`, except in situations where neither of the derived classes is acceptable. For example, exceptions should be thrown by:
> 
> * `ArgumentNullException` whenever `null` is passed to a method that does not accept it as a valid argument.
> 
> * `ArgumentOutOfRangeException` when the value of an argument is outside the range of acceptable values; for example, when the value "46" is passed as the month argument during the creation of a `DateTime`.

This PR is all about best-practices, consistency and code-style, so I would recommend throwing the different exceptions for `null` and empty/white-space values. This also makes debugging a lot easier: just check what exception type was thrown to see what went wrong...

Shazwazza replied:
Ok that's fine but we'll have to log a bug with MS themselves since they are generating code that doesn't adhere to their own standards ;) I think R# can do both exceptions but would need to check. Point is, that these null checks are generally performed by the tooling and that will continue to happen. There's are hundreds of these IsNullOrWhitespace checks with ArgumentException thrown because the tooling provides that. Perhaps the reason for this custom ArgumentNullOrEmptyException was to try to make this check clearer, i dunno.

ronaldbarendse replied:
Visual Studio 2017 (15.9.15) generates the correct checks, so it's probably something in ReSharper or customized settings:
![ArgumentNullException](https://user-images.githubusercontent.com/1051287/63718316-9685d880-c84a-11e9-9454-3aadb693f36d.png)

![ArgumentException](https://user-images.githubusercontent.com/1051287/63718502-072cf500-c84b-11e9-917c-db9d22c13b68.png)

Shazwazza commented on src/Umbraco.Core/Collections/ObservableDictionary.cs:
```diff
@@ -235,15 +236,76 @@ public bool TryGetValue(TKey key, out TValue val)
+                : base("Attempted to insert duplicate key \"" + key + "\" in collection.", innerException)
             {
-                Key = key;
+                this.Key = key;
```
We don't really use `this` coding pattern in this codebase, it's fine but i will remove it.

ronaldbarendse replied:
I'll update this in the next commits 👍

Shazwazza commented:
Ah, i see this is still a draft, all good, have left a review note

nul800sebastiaan commented:
@ronaldbarendse is this still in draft or good to go now? :)

ronaldbarendse commented on src/Umbraco.Core/Exceptions/DataOperationException.cs:
```diff
@@ -1,21 +1,94 @@
+        /// <value>
+        /// The operation.
+        /// </value>
         public T Operation { get; private set; }
```
As `T` could be any type, serialization can fail if the passed in type isn't serializable.

ronaldbarendse replied:
As the exception is currently internal and the only type used for `T` is `MoveOperationStatusType`, we might just remove the generics, right?

ronaldbarendse commented on src/Umbraco.Core/Exceptions/WontImplementException.cs:
```diff
@@ -1,27 +1,52 @@
+    /// reason.
+    /// </remarks>
+    /// <seealso cref="System.NotImplementedException" />
+    [Obsolete("If a method or operation is not, and will not be, implemented, it is invalid, so we should throw an InvalidOperationException instead.")]
```
I've marked `WontImplementException` obsolete, as you should just throw `InvalidOperationException` instead.

ronaldbarendse commented on src/Umbraco.Core/IO/FileSecurityException.cs:
```diff
@@ -1,20 +1,46 @@
+    /// The exception that is thrown when the caller does not have the required permission to access a file.
+    /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Obsolete("Throw an UnauthorizedAccessException instead.")]
```
I've marked `FileSecurityException` obsolete, as it's only used once and just indicates unauthorized access (https://docs.microsoft.com/en-us/dotnet/api/system.unauthorizedaccessexception):
https://github.com/umbraco/Umbraco-CMS/blob/853087a75044b814df458457dc9a1f778cc89749/src/Umbraco.Core/IO/PhysicalFileSystem.cs#L317

ronaldbarendse commented on src/Umbraco.Core/Persistence/EntityNotFoundException.cs:
```diff
@@ -1,38 +1,95 @@
+    /// <seealso cref="System.Exception" />
+    [Obsolete("This exception is only used in MediaController and could be replaced with an HTTP 404 error instead.")]
+    [Serializable]
     public class EntityNotFoundException : Exception
```
I've marked `EntityNotFoundException` obsolete, as it's only used once in an API controller and could be replaced with a generic HTTP 404 error:
https://github.com/umbraco/Umbraco-CMS/blob/56545ad7f1fdee3c2b27e725c89378ccb947f65a/src/Umbraco.Web/Editors/MediaController.cs#L818

ronaldbarendse commented on src/Umbraco.Core/Persistence/EntityNotFoundException.cs:
```diff
@@ -1,38 +1,95 @@
+        /// <value>
+        /// The identifier.
+        /// </value>
         public object Id { get; private set; }
```
Same as in `DataOperationException`, this property can contain a type that isn't serializable.

ronaldbarendse commented on src/Umbraco.Core/Persistence/FaultHandling/RetryLimitExceededException.cs:
```diff
@@ -4,59 +4,51 @@
+        /// Initializes a new instance of the <see cref="RetryLimitExceededException" /> class with a default error message.
         /// </summary>
         public RetryLimitExceededException()
             : this("RetryLimitExceeded")
```
This message could probably be formatted better.

ronaldbarendse commented on src/Umbraco.Web/Install/InstallException.cs:
```diff
@@ -1,44 +1,105 @@
+        /// <value>
+        /// The view model.
+        /// </value>
         public object ViewModel { get; private set; }
```
This property can also contain a non-serializable type.

ronaldbarendse commented:
@Shazwazza Just to summarize the comments:
- Some exceptions can contain property types that may not be serializable (e.g. using `object` or generics): how should we handle this?
- Should we change `MediaController` to `throw new HttpResponseException(HttpStatusCode.NotFound)` instead of the custom exception?

Removing usage of obsoleted exceptions should probably be done in a seperate PR, as that might cause breaking changes, right?

Shazwazza commented:
Hi @ronaldbarendse 

> Some exceptions can contain property types that may not be serializable (e.g. using object or generics): how should we handle this?

I'm not really sure, if there's no clear cut way to handle this then should we just not allow serialization? It hasn't been a problem to date afaik so would it make sense to just not support the serialization for those types?

> Should we change MediaController to throw new HttpResponseException(HttpStatusCode.NotFound) instead of the custom exception?

yep i think so!

> Removing usage of obsoleted exceptions should probably be done in a seperate PR, as that might cause breaking changes, right?

I think this depends on where the usages are. If they are in our back office controllers, we are allowed those types of changes, same with at the service/repo level and some other places too.

ronaldbarendse commented on src/Umbraco.Core/Collections/ObservableDictionary.cs:
```diff
@@ -235,15 +236,76 @@ public bool TryGetValue(TKey key, out TValue val)
+        /// <seealso cref="System.Collections.Generic.IReadOnlyDictionary{TKey, TValue}" />
+        /// <seealso cref="System.Collections.Generic.IDictionary{TKey, TValue}" />
+        [Serializable]
         internal class DuplicateKeyException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse replied:
This exception also seems redundant, as it's only used in its parent class (`ObservableDictionary<TKey, TValue>`) and can be replaced with a `ArgumentException` (like https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.add).

ronaldbarendse replied:
The `InvalidOperationException` thrown by `ObservableDictionary<TKey, TValue>` when the key isn't found should also be changed to a `KeyNotFoundException` (like https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.item).

ronaldbarendse commented on src/Umbraco.Core/Composing/TypeLoader.cs:
```diff
@@ -813,11 +814,44 @@ public void Add(Type type)
         /// </summary>
+        /// <seealso cref="System.Exception" />
+        [Serializable]
         internal class CachedTypeNotFoundInFileException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Core/Exceptions/DataOperationException.cs:
```diff
@@ -1,21 +1,94 @@
+    /// <typeparam name="T"></typeparam>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class DataOperationException<T> : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Core/Migrations/DataLossException.cs:
```diff
@@ -1,22 +1,45 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class DataLossException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Web/Media/Exif/ExifExceptions.cs:
```diff
@@ -1,42 +1,87 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class NotValidExifFileException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Web/Media/Exif/ExifExceptions.cs:
```diff
@@ -1,42 +1,87 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class UnknownEnumTypeException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Web/Media/Exif/JPEGExceptions.cs:
```diff
@@ -1,99 +1,211 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class NotValidImageFileException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Web/Media/Exif/JPEGExceptions.cs:
```diff
@@ -1,99 +1,211 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class NotValidJPEGFileException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Web/Media/Exif/JPEGExceptions.cs:
```diff
@@ -1,99 +1,211 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class NotValidTIFFileException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Web/Media/Exif/JPEGExceptions.cs:
```diff
@@ -1,99 +1,211 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class NotValidTIFFHeader : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented on src/Umbraco.Web/Media/Exif/JPEGExceptions.cs:
```diff
@@ -1,99 +1,211 @@
     /// </summary>
+    /// <seealso cref="System.Exception" />
+    [Serializable]
     internal class SectionExceeds64KBException : Exception
```
All exceptions should be `public`, as consumers can't catch it otherwise!

ronaldbarendse commented:
I've done some more cleaning up, so the following exceptions are now removed:
- `EntityNotFoundException`: used once in `MediaController` and replaced with an `HttpResponseException`
- `FileSecurityException`: used once in `PhysicalFileSystem` and replaced with an `UnauthorizedAccessException`
- `WontImplementException`: replaced with an `InvalidOperationException` including a descriptive message

A lot of the exceptions are also `internal` and thus not able to get catched by consumers (e.g. `try {} catch (DataOperationException) {}` is not possible). This might not be a problem it the exception is _always_ cached by an internal method (and not rethrown), but ensuring that might be quite hard...

ronaldbarendse commented:
I'll redo some of the commits, so it only obsoletes custom unneeded/incompatible (non-serializable) exceptions and removes all usages of them...

Returning different exceptions might cause breaking changes, especially if the base class is different, but I don't think that would be a real issue (might be good to merge this into the next minor version though).

The obsoleted `internal` exceptions _can_ get removed, as those can't be catched anyway. All other should be changed to `public`: https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1064-exceptions-should-be-public.

I'll make sure the changes are done in separate commits, so you can always cherry-pick what gets merged 👍

ronaldbarendse commented:
I've rewritten this PR, so it doesn't remove any of the custom public exceptions (causing a binary breaking change). We should continue the conversation on PR https://github.com/umbraco/Umbraco-CMS/pull/6599.
